x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java

```diff
@@ -0,0 +1,209 @@
+        SearchAction.NAME,
+        Set.of(
+            SearchTransportService.FREE_CONTEXT_ACTION_NAME,
+            SearchTransportService.DFS_ACTION_NAME,
+            SearchTransportService.QUERY_ACTION_NAME,
+            SearchTransportService.QUERY_ID_ACTION_NAME,
+            SearchTransportService.FETCH_ID_ACTION_NAME,
+            SearchTransportService.QUERY_CAN_MATCH_NAME,
+            SearchTransportService.QUERY_CAN_MATCH_NODE_NAME
+        )
+    );
+
+    /**
+     * This method sets {@link ParentActionAuthorization} as a header in the thread context,
+     * which will be used for skipping authorization of child actions if the following conditions are met:
+     *
+     * <ul>
+     * <li>parent action is one of the white listed in {@link #CHILD_ACTIONS_PRE_AUTHORIZED_BY_PARENT}</li>
+     * <li>FLS and DLS are not configured</li>
+     * <li>RBACEngine was used to authorize parent request and not a custom authorization engine</li>
+     * </ul>
+     */
+    public static void maybeSkipChildrenActionAuthorization(
+        SecurityContext securityContext,
+        TransportRequest parentRequest,
+        AuthorizationContext parentAuthorizationContext
+    ) {
+        final String parentAction = parentAuthorizationContext.getAction();
+        if (CHILD_ACTIONS_PRE_AUTHORIZED_BY_PARENT.containsKey(parentAction) == false) {
+            // This is not one of the white listed parent actions.
+            return;
+        }
+
+        final IndicesAccessControl indicesAccessControl = parentAuthorizationContext.getIndicesAccessControl();
+        if (indicesAccessControl == null) {
+            // This can happen if the parent request was authorized by index name only - e.g. bulk request
+            // A missing IAC is not an error, but it means we can't safely tie authz of the child action to the parent authz
+            return;
+        }
+
+        // Just a sanity check. If we ended up here, the authz should have been granted.
+        if (indicesAccessControl.isGranted() == false) {
+            return;
+        }
+
+        if (parentRequest instanceof IndicesRequest == false) {
+            // We can only pre-authorize indices request.
+            return;
+        }
+
+        final Role role = RBACEngine.maybeGetRBACEngineRole(parentAuthorizationContext.getAuthorizationInfo());
+        if (role == null) {
+            // If role is null, it means a custom authorization engine is in use, hence we cannot do the optimization here.
+            return;
+        }
+
+        // We can't safely pre-authorize actions if DLS or FLS is configured without passing IAC as well with the authorization result.
+        // For simplicity, we only pre-authorize actions when FLS and DLS are not configured, otherwise we would have to compute and send
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
```

Comment:
I think this is too general.
The role can have DLS/FLS permissions defined but which don't cover the indices in the request.
Something like `indicesAccessControl.getFieldAndDocumentLevelSecurityUsage().hasFieldOrDocumentLevelSecurity()` would be more precise.

Reply:
That makes sense, but going with the approach to check if `indicesAccessControl` has DLS or FLS would mean that we have to check access for all indices. Since we plan to make IAC computed lazily, it would mean that we would always access IAC on a coordinating node and we want to avoid that.

Reply:
Yeah, that's a good point. Okay lets keep it as is for now.